pkg/dockerregistry/server/quotarestrictedblobstore.go

```diff
@@ -122,7 +123,51 @@ func (bw *quotaRestrictedBlobWriter) Commit(ctx context.Context, provisional dis
 
 // admitBlobWrite checks whether the blob does not exceed image quota, if set. Returns
 // ErrAccessDenied error if the quota is exceeded.
-func admitBlobWrite(ctx context.Context, repo *repository) error {
+func admitBlobWrite(ctx context.Context, repo *repository, size int64) error {
+	if size < 1 {
+		return nil
+	}
+
+	err := admitLimitRanges(ctx, repo, size)
+	if err != nil {
+		return err
+	}
+
+	err = admitQuotas(ctx, repo)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// admitLimitRanges checks the blob against any established limit ranges.
+func admitLimitRanges(ctx context.Context, repo *repository, size int64) error {
+	limitranges, err := repo.limitClient.LimitRanges(repo.namespace).List(kapi.ListOptions{})
+	if err != nil {
+		if kerrors.IsForbidden(err) {
+			context.GetLogger(ctx).Warnf("Cannot list limitranges because of outdated cluster roles: %v", err)
```

Comment:
If Quota is opt in on the registry, then yes, you can hard error here.  It
will eventually be valid to run read-only registries - I may choose not to
expose to them access to read all usage information in the cluster.

My other comment is that we shouldn't even be doing this here, but I'd like
to hear the reasoning.

On Tue, Mar 22, 2016 at 9:00 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/dockerregistry/server/quotarestrictedblobstore.go
> <https://github.com/openshift/origin/pull/7989#discussion_r57095634>:
>
> > +	}
> > +
> > +	err = admitQuotas(ctx, repo)
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	return nil
> > +}
> > +
> > +// admitLimitRanges checks the blob against any established limit ranges.
> > +func admitLimitRanges(ctx context.Context, repo *repository, size int64) error {
> > +	limitranges, err := repo.limitClient.LimitRanges(repo.namespace).List(kapi.ListOptions{})
> > +	if err != nil {
> > +		if kerrors.IsForbidden(err) {
> > +			context.GetLogger(ctx).Warnf("Cannot list limitranges because of outdated cluster roles: %v", err)
>
> A forbidden error in response to an API request is not a strong enough
> signal to disable this feature on. That means a future role modification
> error could silently open up the registry to be unquotaed
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7989/files/aee1f1537969521a96e027cec8461be7689d4436..e8e272d1d258f7b0f5941504ca83cb792bfc2098#r57095634>
>